File: src/themelad.py
import os
import sys


def get_theme_paths() -> list[str]:
    current_dir = os.getcwd()
    args = sys.argv[1:]
    if len(args):
        theme_paths = []
        for theme_path in args:
            full_path = os.path.join(current_dir, theme_path)
            if os.path.isdir(full_path):
                theme_paths.append(full_path)
        return theme_paths
    else:
        theme_paths = []
        dirs = os.listdir(current_dir)
        for dir in dirs:
            theme_paths.append(os.path.join(current_dir, dir))
        return theme_paths


if __name__ == '__main__':
    theme_paths = get_theme_paths()
    print(theme_paths)

File: src/themelad.nim
import os
import sequtils
import strformat
import sugar


proc getThemePaths(): seq[string] =
  let currentDir = getCurrentDir()
  if paramCount() > 0:
    let args = commandLineParams()
    return args.map(arg => joinPath(currentDir, arg))
  else:
    let dirs = joinPath(currentDir, "/*")
    for dir in walkDirs dirs:
      result.add(dir)


func getMasterThemeName(masterThemePath: string): string =
  let themeName = lastPathPart masterThemePath
  fmt"{themeName}-master"


func getMasterThemePath(themeDir: string): string =
  let masterThemeName = getMasterThemeName themeDir
  joinPath(themeDir, masterThemeName)


func hasMasterTheme(themeDir: string): bool =
  let masterThemeDir = getMasterThemePath themeDir
  dirExists masterThemeDir


func getValidThemePaths(themeDirs: seq[string]): seq[string] =
  themeDirs
    .filter(dirExists)
    .filter(hasMasterTheme)


when isMainModule:
  let
    themePaths = getThemePaths()
    validThemePaths = getValidThemePaths themePaths
  
  echo validThemePaths
